src/bsoncxx/enums/type.hpp

```diff
@@ -17,7 +17,7 @@
 #endif
 
 BSONCXX_ENUM(double, 0x01)
-BSONCXX_ENUM(utf8, 0x02)
+BSONCXX_ENUM(string, 0x02)
```

Comment:
I believe this renaming causes some API removal. 

grepping through all .hpp files for `#include <bsoncxx/enums/type.hpp>`, I think these would be removed.

1. `type::k_utf8` is defined in `src/bsoncxx/types.hpp`.
2. `error_code::k_need_element_type_k_utf8` and `error_code::k_cannot_append_utf8` are defined. These are integer error codes which I don't believe we can change.

Both cases are enum values.

For those cases, we may need to redefine each at the same integer value.

Ideally we can deprecate the old utf8 value. There may not be a completely portable way of doing so, but `#pragma deprecated` would at least deprecate for compilers that recognize the pragma. Ideally, users are informed by a compiler warning. But at the very least, we can give them notice via release notes.

Here's what I'm thinking.

```
///
/// An enumeration of each BSON type.
/// These x-macros will expand to be of the form:
///    k_double = 0x01,
///    k_string = 0x02,
///    k_document = 0x03,
///    k_array = 0x04 ...
///
enum class type : std::uint8_t {
#define BSONCXX_ENUM(name, val) k_##name = val,
#include <bsoncxx/enums/type.hpp>
#undef BSONCXX_ENUM
    k_utf8 = 0x02
};

#pragma deprecated(k_utf8)
```

I think this could be a motivation for doing CXX-641 before the next release as well. That way we can double check ABI changes before we release. I moved that to "Needs Triage".

Reply:
Done. I added `k_utf8` and deprecated it using `BSONCXX_DEPRECATED`. For the error codes, I added them back to `error_code.hpp`. They're at the end of the enum to avoid changing other error code values. I set them to the `_string` variants since they share the same values. I did _not_ add them to the switch statements in `error_code.cpp`. You can't add a new case for the same values.

Reply:
So it appears that the ability deprecated enum values didn't appear until GCC-6 ([source](https://gcc.gnu.org/gcc-6/changes.html)):
 
> The C and C++ compilers now support attributes on enumerators. For instance, it is now possible to mark enumerators as deprecated:
> 
> 
>
> ```
> enum {
>   newval,
>   oldval __attribute__ ((deprecated ("too old")))
> };
> ```

The code snippet above (`#pragma deprecated(k_utf8)`) doesn't even compile locally so I'm not sure if it's possible to deprecate this.

Reply:
We might just have to leave this and error code as is until the next major release. @samantharitter any ideas?